source/agora/node/admin/QRCodeInterface.d

```diff
@@ -0,0 +1,269 @@
+            )
+        );
+
+        login_info.voter_card.signature =
+            this.key_pair.secret.sign(hashFull(login_info)[]);
+
+        res.headers["Content-Type"] = "image/svg+xml";
+        res.headers["Vary"] = "Accept-Encoding";
+        res.writeBody(createQRcode(login_info));
+    }
+
+    /***************************************************************************
+
+        Respond with QR code containing EncryptionKey.
+
+        Enter the name of the APP & height where the encryptionKey will be used
+        as the get query.
+
+        The encryption key is hashed using the blake2 cryptographic
+        hash function.
+        The encryption key is derived from the pre-image as follows:
+        Hash(pre_image, "APPNAME");
+
+    ***************************************************************************/
+
+    public void keyQRCreate (scope HTTPServerRequest req,
+        scope HTTPServerResponse res)
+    {
+        enforceHTTP("app" in req.query && "height" in req.query,
+            HTTPStatus.badRequest, "Missing app/height field.");
+
+        auto app = req.query["app"];
+        auto height = Height.fromString(req.query["height"]);
+
+        // TODO: PreImage to be released at that height
+        // Or it can be pre-released, so it can be `height + PreimageRevealPeriod`
+        PreImageInfo pre_image;
+
+        // blake2b hashing
+        const Hash key = hashMulti(pre_image, app);
+
+        EncryptionKey encryptionKey = EncryptionKey(
+            app,
+            height,
+            key
+        );
+
+        res.headers["Content-Type"] = "image/svg+xml";
+        res.headers["Vary"] = "Accept-Encoding";
+        res.writeBody(createQRcode(encryptionKey));
+    }
+
+    /***************************************************************************
+
+        SVG format QR code generator
+
+        Use barcode library to make QR code of JSON format.
+        Convert the data structure to JSON and return the QR code image
+        of the SVG format.
+
+        Params:
+            t = Type of struct to QR code
+
+        Returns:
+            QR code of XML string
+
+    ***************************************************************************/
+
+    private string createQRcode (T)(T t)
+        if (is(T == struct))
+    {
+        auto str = serializeToJsonString(t);
+
+        auto svgDrawer = new BaseBarCodeSvgDrawer;
+        svgDrawer.fixSizeMode = true;
+        svgDrawer.W = 400;
+        svgDrawer.H = 400;
+
+        auto qr = new Qr;
+        auto bc = qr.encode(str);
+        return(svgDrawer.draw(bc));
+    }
+}
+
+/*******************************************************************************
+
+    Defines the LoginInfo.
+
+*******************************************************************************/
+
+public struct LoginInfo
+{
+    // The randomly generated temporary private key encoded as string
+    public string private_key;
+
+    public VoterCard voter_card;
+
+}
```

Comment:
Need a bit of documentation / style fix.

Reply:
Added document and removed a line.